src/Data/PQueue/Prio/Max.hs

```diff
@@ -172,8 +176,15 @@ instance Ord k => Foldable (MaxPQueue k) where
   foldr f z (MaxPQ q) = foldr f z q
   foldl f z (MaxPQ q) = foldl f z q
 
+  length = size
+  null = null
+
+-- | Traverses in descending order. 'mapM' is strictly accumulating like
+-- 'mapMWithKey'.
 instance Ord k => Traversable (MaxPQueue k) where
   traverse f (MaxPQ q) = MaxPQ <$> traverse f q
+  mapM = mapMWithKey . const
+  sequence = mapM id
```

Comment:
Is that really lawful for non-strict monads?

Reply:
Non-strict monads aren't lawful themselves. Consider lazy `State`.

```haskell
newtype State s a = State {runState :: s -> (a, s)}
instance Functor (State s) where
  -- This is equivalent to liftM
  fmap f (State st) = State $ \s ->
    case st s of
      ~(a, s') -> (f a, s')
```

This violates the functor identity law: `fmap id m = m`.

```haskell
runStateT (fmap id undefined) ()
=
case undefined () of
  ~(a, s') -> (id a, s')
=
(undefined, undefined)
```

But

```haskell
runStateT undefined () = undefined
```

There's always a certain amount of wishy-washiness when dealing with non-strict monads. I think it's perfectly fine to lean on "fast and loose reasoning is morally correct" and go with it. Note also that when you "run" a non-strict monad, you're almost always doing so *because you want a result*. But `traverseWithKey` won't give you a result when `mapMWithKey` doesn't.